File: python/prism_sdk/autonomous_http_snapshot_store.py
"""Bounded HTTP persistence for metadata-only JSON snapshot text stores.

The store is intentionally schema-neutral.  Learning, run traces, evaluator state, goals, and
queues validate their own JSON before this adapter sees it.  This layer contributes endpoint
admission, bounded request/response handling, transient header resolution, no redirects, timeout
classification, and conditional-write fencing; it does not claim database durability or
distributed consensus.
"""

from __future__ import annotations

import ipaddress
import json
from typing import Any, Callable, Mapping, Sequence
from urllib.error import HTTPError, URLError
from urllib.parse import urlsplit
from urllib.request import HTTPRedirectHandler, Request, build_opener

from .authoring import canonical_json
from .errors import ArgumentError, TransportError


AUTONOMOUS_HTTP_SNAPSHOT_STORE_SCHEMA = "bioprism-python-autonomous-http-snapshot-store/0.1"
MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESOURCE_BYTES = 512
MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_REQUEST_BYTES = 4_000_000
MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESPONSE_BYTES = 4_000_000
MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_TIMEOUT_SECONDS = 120.0
MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_COUNT = 64
MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_BYTES = 8_192


class _NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, *_args: Any, **_kwargs: Any) -> None:
        return None


def _bounded_text(name: str, value: Any, maximum: int) -> str:
    if not isinstance(value, str) or not value.strip() or "\x00" in value or len(value.encode("utf-8")) > maximum:
        raise ArgumentError(f"{name} is outside its bounded text contract")
    return value


def _loopback(host: str) -> bool:
    normalized = host.strip("[]").lower()
    if normalized == "localhost":
        return True
    try:
        return ipaddress.ip_address(normalized).is_loopback
    except ValueError:
        return False


def _host_allowed(host: str, allowed_hosts: tuple[str, ...], allow_loopback: bool) -> bool:
    normalized = host.strip("[]").lower()
    if allow_loopback and _loopback(normalized):
        return True
    return any(
        (allowed.startswith("*.") and normalized.endswith(allowed[1:]) and normalized != allowed[2:])
        or normalized == allowed
        for allowed in allowed_hosts
    )


def _validate_snapshot_text(value: str, maximum: int) -> str:
    if not isinstance(value, str) or len(value.encode("utf-8")) > maximum:
        raise ArgumentError("HTTP snapshot store snapshot exceeds its request bound")
    try:
        parsed = json.loads(value)
    except json.JSONDecodeError as error:
        raise ArgumentError("HTTP snapshot store snapshot must be valid JSON") from error
    if not isinstance(parsed, dict):
        raise ArgumentError("HTTP snapshot store snapshot must be a JSON object")
    if canonical_json(parsed) != value:
        raise ArgumentError("HTTP snapshot store snapshot must use canonical JSON")
    return value


def _headers(value: Mapping[str, str]) -> dict[str, str]:
    if not isinstance(value, Mapping) or len(value) > MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_COUNT:
        raise ArgumentError("HTTP snapshot store headers are outside their bound")
    result: dict[str, str] = {}
    for raw_name, raw_value in value.items():
        name = _bounded_text("HTTP snapshot store header name", raw_name, 256)
        if any(character.isspace() for character in name) or ":" in name or "\r" in name or "\n" in name:
            raise ArgumentError("HTTP snapshot store header name is unsafe")
        text = _bounded_text("HTTP snapshot store header value", raw_value, MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_BYTES)
        if "\r" in text or "\n" in text:
            raise ArgumentError("HTTP snapshot store header value is unsafe")
        for existing in tuple(result):
            if existing.lower() == name.lower():
                del result[existing]
        result[name] = text
    return result


class AutonomousHttpSnapshotTextStore:
    """A caller-owned remote text store with atomic conditional PUT support."""

    def __init__(
        self,
        endpoint: str,
        resource: str,
        *,
        allowed_hosts: Sequence[str] = (),
        require_https: bool = True,
        allow_loopback: bool = False,
        timeout_seconds: float = 30.0,
        max_request_bytes: int = MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_REQUEST_BYTES,
        max_response_bytes: int = MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESPONSE_BYTES,
        header_resolver: Callable[[Mapping[str, Any]], Mapping[str, str]] | None = None,
        opener: Callable[[Request, float], Any] | None = None,
    ) -> None:
        if not isinstance(require_https, bool) or not isinstance(allow_loopback, bool):
            raise ArgumentError("HTTP snapshot store HTTPS policy must be boolean")
        endpoint_text = _bounded_text("HTTP snapshot store endpoint", endpoint, 8_192)
        try:
            parsed = urlsplit(endpoint_text)
        except ValueError as error:
            raise ArgumentError("HTTP snapshot store endpoint is not a valid URL") from error
        if parsed.scheme not in {"http", "https"} or not parsed.hostname or parsed.username or parsed.password or parsed.fragment:
            raise ArgumentError("HTTP snapshot store endpoint failed URL admission")
        if require_https and parsed.scheme != "https" and not (allow_loopback and parsed.scheme == "http" and _loopback(parsed.hostname)):
            raise ArgumentError("HTTP snapshot store endpoint must use HTTPS unless loopback development is enabled")
        if not isinstance(allowed_hosts, Sequence) or isinstance(allowed_hosts, (str, bytes)) or len(allowed_hosts) > 128:
            raise ArgumentError("HTTP snapshot store allowed_hosts are outside their bound")
        normalized_hosts: list[str] = []
        for host in allowed_hosts:
            normalized = _bounded_text("HTTP snapshot store allowed host", host, 512).lower().strip("[]")
            if "://" in normalized or "/" in normalized or "@" in normalized or (normalized.startswith("*.") and len(normalized) <= 2):
                raise ArgumentError("HTTP snapshot store allowed host is malformed")
            normalized_hosts.append(normalized)
        if not _host_allowed(parsed.hostname, tuple(normalized_hosts), allow_loopback):
            raise ArgumentError("HTTP snapshot store endpoint host is outside its allowlist")
        self.endpoint = endpoint_text
        self.resource = _bounded_text("HTTP snapshot store resource", resource, MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESOURCE_BYTES)
        if not all(character.isalnum() or character in "_.:/+-" for character in self.resource):
            raise ArgumentError("HTTP snapshot store resource contains unsafe characters")
        if isinstance(timeout_seconds, bool) or not isinstance(timeout_seconds, (int, float)) or not 0.1 <= float(timeout_seconds) <= MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_TIMEOUT_SECONDS:
            raise ArgumentError("HTTP snapshot store timeout_seconds is outside its bound")
        for name, value, maximum in (
            ("max_request_bytes", max_request_bytes, MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_REQUEST_BYTES),
            ("max_response_bytes", max_response_bytes, MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESPONSE_BYTES),
        ):
            if isinstance(value, bool) or not isinstance(value, int) or not 1 <= value <= maximum:
                raise ArgumentError(f"HTTP snapshot store {name} is outside its bound")
        if header_resolver is not None and not callable(header_resolver):
            raise ArgumentError("HTTP snapshot store header_resolver must be callable")
        self.require_https = require_https
        self.allow_loopback = allow_loopback
        self.timeout_seconds = float(timeout_seconds)
        self.max_request_bytes = max_request_bytes
        self.max_response_bytes = max_response_bytes
        self._allowed_hosts = tuple(normalized_hosts)
        self.header_resolver = header_resolver
        self._opener = opener or build_opener(_NoRedirect()).open
        if not callable(self._opener):
            raise ArgumentError("HTTP snapshot store opener must be callable")

    def describe(self) -> dict[str, Any]:
        parsed = urlsplit(self.endpoint)
        return {
            "schema": AUTONOMOUS_HTTP_SNAPSHOT_STORE_SCHEMA,
            "resource": self.resource,
            "host": parsed.hostname,
            "scheme": parsed.scheme,
            "require_https": self.require_https,
            "allow_loopback": self.allow_loopback,
            "timeout_seconds": self.timeout_seconds,
            "max_request_bytes": self.max_request_bytes,
            "max_response_bytes": self.max_response_bytes,
            "cas": "if_match_digest_or_if_none_match_star",
            "credentials": "transient_header_resolver;never_returned",
            "retention": "metadata_only;caller_schema_validation_required",
            "secret_material": "never_returned",
        }

    def read(self) -> str | None:
        status, body = self._request("read", "GET", None, None)
        if status == 404:
            return None
        if not 200 <= status < 300:
            raise TransportError(f"HTTP snapshot store read returned status {status}")
        return _validate_snapshot_text(body, self.max_response_bytes)

    def write(self, value: str) -> None:
        encoded = _validate_snapshot_text(value, self.max_request_bytes)
        status, _body = self._request("write", "PUT", encoded, None)
        if not 200 <= status < 300:
            raise TransportError(f"HTTP snapshot store write returned status {status}")

    def write_if_unchanged(self, expected_snapshot_digest: str | None, value: str) -> bool:
        if expected_snapshot_digest is not None and (not isinstance(expected_snapshot_digest, str) or len(expected_snapshot_digest) != 64 or any(character not in "0123456789abcdef" for character in expected_snapshot_digest)):
            raise ArgumentError("HTTP snapshot store expected snapshot digest must be a lowercase SHA-256 digest or null")
        encoded = _validate_snapshot_text(value, self.max_request_bytes)
        status, _body = self._request("write_if_unchanged", "PUT", encoded, expected_snapshot_digest)
        if status in {409, 412}:
            return False
        if not 200 <= status < 300:
            raise TransportError(f"HTTP snapshot store write_if_unchanged returned status {status}")
        return True

    def _request(self, operation: str, method: str, body: str | None, expected_snapshot_digest: str | None) -> tuple[int, str]:
        context = {
            "operation": operation,
            "resource": self.resource,
            "expected_snapshot_digest": expected_snapshot_digest,
        }
        resolved: Mapping[str, str] = {}
        if self.header_resolver is not None:
            resolved = self.header_resolver(context)
        request_headers = _headers(resolved)
        request_headers["Accept"] = "application/json"
        request_headers["X-Aurora-Snapshot-Resource"] = self.resource
        data = None
        if body is not None:
            data = body.encode("utf-8")
            if len(data) > self.max_request_bytes:
                raise ArgumentError("HTTP snapshot store request exceeds its byte bound")
            request_headers["Content-Type"] = "application/json"
        if operation == "write_if_unchanged":
            request_headers["If-None-Match" if expected_snapshot_digest is None else "If-Match"] = "*" if expected_snapshot_digest is None else f'"{expected_snapshot_digest}"'
        request = Request(self.endpoint, data=data, headers=request_headers, method=method)
        try:
            response = self._opener(request, self.timeout_seconds)
            with response:
                raw_status = getattr(response, "status", None)
                status = int(raw_status if raw_status is not None else response.getcode())
                return status, self._read_response(response)
        except HTTPError as error:
            return int(error.code), ""
        except (TimeoutError, URLError, OSError) as error:
            raise TransportError(f"HTTP snapshot store {operation} transport failed") from error

    def _read_response(self, response: Any) -> str:
        chunks: list[bytes] = []
        total = 0
        while True:
            chunk = response.read(min(64 * 1024, self.max_response_bytes + 1 - total))
            if not chunk:
                break
            chunks.append(chunk)
            total += len(chunk)
            if total > self.max_response_bytes:
                raise TransportError("HTTP snapshot store response exceeded its byte bound")
        try:
            return b"".join(chunks).decode("utf-8")
        except UnicodeDecodeError as error:
            raise TransportError("HTTP snapshot store response was not valid UTF-8") from error


__all__ = [
    "AUTONOMOUS_HTTP_SNAPSHOT_STORE_SCHEMA",
    "MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESOURCE_BYTES",
    "MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_REQUEST_BYTES",
    "MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESPONSE_BYTES",
    "MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_TIMEOUT_SECONDS",
    "MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_COUNT",
    "MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_BYTES",
    "AutonomousHttpSnapshotTextStore",
]

File: typescript/src/autonomous-http-snapshot-store.ts
import { ArgumentError, ProtocolError, ResponseTooLargeError, TransportError, isObject } from "./errors.js";
import { canonicalJson } from "./tooling.js";
import type { JsonObject } from "./types.js";

/**
 * Provider-neutral HTTP transport for the SDK's metadata-only JSON text stores.
 *
 * The store intentionally knows nothing about jobs, goals, learning, or evidence. Those higher
 * layers validate their own schemas before handing canonical JSON here. This adapter owns only
 * bounded transport, endpoint policy, timeout/cancellation, and compare-and-swap headers. A
 * caller-supplied header resolver may close over a short-lived auth session; the resolver output
 * is used for one request and is never returned, persisted, or included in an error.
 */
export const AUTONOMOUS_HTTP_SNAPSHOT_STORE_SCHEMA = "bioprism-typescript-autonomous-http-snapshot-store/0.1" as const;
export const MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESOURCE_BYTES = 512;
export const MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_REQUEST_BYTES = 4_000_000;
export const MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESPONSE_BYTES = 4_000_000;
export const MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_TIMEOUT_MS = 120_000;
export const MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_COUNT = 64;
export const MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_BYTES = 8_192;

const DEFAULT_TIMEOUT_MS = 30_000;
const SAFE_METHODS = ["GET", "PUT"] as const;
const OPERATIONS = ["read", "write", "write_if_unchanged"] as const;

export type AutonomousHttpSnapshotStoreOperation = typeof OPERATIONS[number];

export interface AutonomousHttpSnapshotStoreHeaderContext extends JsonObject {
  operation: AutonomousHttpSnapshotStoreOperation;
  resource: string;
  expected_snapshot_digest: string | null;
}

export type AutonomousHttpSnapshotStoreHeaderResolver = (
  context: AutonomousHttpSnapshotStoreHeaderContext,
) => Record<string, string> | Promise<Record<string, string>>;

export interface AutonomousHttpSnapshotStorePolicy {
  allowedHosts?: readonly string[];
  requireHttps?: boolean;
  allowLoopback?: boolean;
  timeoutMs?: number;
  maxRequestBytes?: number;
  maxResponseBytes?: number;
}

export interface AutonomousHttpSnapshotStoreOptions extends AutonomousHttpSnapshotStorePolicy {
  endpoint: string | URL;
  resource: string;
  fetch?: AutonomousHttpSnapshotStoreFetch;
  headerResolver?: AutonomousHttpSnapshotStoreHeaderResolver;
  signal?: AbortSignal;
}

export type AutonomousHttpSnapshotStoreFetch = (input: string | URL, init?: RequestInit) => Promise<Response>;

export interface AutonomousHttpSnapshotStoreDescription extends JsonObject {
  schema: typeof AUTONOMOUS_HTTP_SNAPSHOT_STORE_SCHEMA;
  resource: string;
  host: string;
  scheme: "http" | "https";
  require_https: boolean;
  allow_loopback: boolean;
  timeout_ms: number;
  max_request_bytes: number;
  max_response_bytes: number;
  cas: "if_match_digest_or_if_none_match_star";
  credentials: "transient_header_resolver;never_returned";
  retention: "metadata_only;caller_schema_validation_required";
  secret_material: "never_returned";
}

function boundedText(name: string, value: unknown, maximum: number): string {
  if (typeof value !== "string" || !value.trim() || value.includes("\u0000") || new TextEncoder().encode(value).byteLength > maximum) {
    throw new ArgumentError(`${name} is outside its bounded text contract`);
  }
  return value;
}

function isLoopback(host: string): boolean {
  const normalized = host.toLowerCase().replace(/^\[/, "").replace(/\]$/, "");
  if (normalized === "localhost" || normalized === "::1") return true;
  const octets = normalized.split(".");
  return octets.length === 4 && octets[0] === "127" && octets.slice(1).every((part) => /^\d+$/.test(part) && Number(part) <= 255);
}

function normalizeHost(host: string): string {
  return boundedText("HTTP snapshot store allowed host", host, 512).toLowerCase().replace(/^\[/, "").replace(/\]$/, "");
}

function hostAllowed(host: string, allowedHosts: readonly string[], allowLoopback: boolean): boolean {
  const normalized = host.toLowerCase().replace(/^\[/, "").replace(/\]$/, "");
  if (allowLoopback && isLoopback(normalized)) return true;
  return allowedHosts.some((allowed) => allowed.startsWith("*.")
    ? normalized.endsWith(allowed.slice(1)) && normalized !== allowed.slice(2)
    : normalized === allowed);
}

function validateEndpoint(endpoint: string | URL, allowedHosts: readonly string[], requireHttps: boolean, allowLoopback: boolean): URL {
  let parsed: URL;
  try {
    parsed = endpoint instanceof URL ? new URL(endpoint.toString()) : new URL(boundedText("HTTP snapshot store endpoint", endpoint, 8_192));
  } catch {
    throw new ArgumentError("HTTP snapshot store endpoint is not a valid URL");
  }
  if (parsed.protocol !== "https:" && parsed.protocol !== "http:") throw new ArgumentError("HTTP snapshot store endpoint must use HTTP or HTTPS");
  if (requireHttps && parsed.protocol !== "https:" && !(allowLoopback && parsed.protocol === "http:" && isLoopback(parsed.hostname))) throw new ArgumentError("HTTP snapshot store endpoint must use HTTPS unless loopback development is explicitly enabled");
  if (parsed.username || parsed.password || parsed.hash) throw new ArgumentError("HTTP snapshot store endpoint cannot contain credentials or a fragment");
  if (!hostAllowed(parsed.hostname, allowedHosts, allowLoopback)) throw new ArgumentError("HTTP snapshot store endpoint host is outside its allow-list");
  return parsed;
}

function normalizeHeaders(value: Record<string, string>): Record<string, string> {
  if (!value || typeof value !== "object" || Array.isArray(value) || Object.keys(value).length > MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_COUNT) throw new ArgumentError("HTTP snapshot store headers are outside their bound");
  const normalized: Record<string, string> = {};
  for (const [rawName, rawValue] of Object.entries(value)) {
    const name = boundedText("HTTP snapshot store header name", rawName, 256);
    if (/\s|:|\r|\n/.test(name)) throw new ArgumentError("HTTP snapshot store header name is unsafe");
    const text = boundedText("HTTP snapshot store header value", rawValue, MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_HEADER_BYTES);
    if (/\r|\n/.test(text)) throw new ArgumentError("HTTP snapshot store header value is unsafe");
    for (const existing of Object.keys(normalized)) if (existing.toLowerCase() === name.toLowerCase()) delete normalized[existing];
    normalized[name] = text;
  }
  return normalized;
}

function assertDigest(name: string, value: string | null): void {
  if (value !== null && !/^[0-9a-f]{64}$/.test(value)) throw new ArgumentError(`${name} must be a lowercase SHA-256 digest or null`);
}

async function boundedResponseText(response: Response, maximum: number): Promise<string> {
  const declared = response.headers.get("content-length");
  if (declared !== null && /^\d+$/.test(declared) && Number(declared) > maximum) throw new ResponseTooLargeError(maximum);
  if (response.body && typeof response.body.getReader === "function") {
    const reader = response.body.getReader();
    const chunks: Uint8Array[] = [];
    let total = 0;
    try {
      while (true) {
        const next = await reader.read();
        if (next.done) break;
        const chunk = next.value instanceof Uint8Array ? next.value : new Uint8Array(next.value);
        total += chunk.byteLength;
        if (total > maximum) {
          await reader.cancel();
          throw new ResponseTooLargeError(maximum);
        }
        chunks.push(chunk);
      }
    } catch (error) {
      try { await reader.cancel(); } catch { /* preserve the bounded transport failure */ }
      throw error;
    }
    const bytes = new Uint8Array(total);
    let offset = 0;
    for (const chunk of chunks) {
      bytes.set(chunk, offset);
      offset += chunk.byteLength;
    }
    return new TextDecoder().decode(bytes);
  }
  const text = await response.text();
  if (new TextEncoder().encode(text).byteLength > maximum) throw new ResponseTooLargeError(maximum);
  return text;
}

function validateSnapshotText(value: string, maximum: number): string {
  if (typeof value !== "string" || new TextEncoder().encode(value).byteLength > maximum) throw new ArgumentError("HTTP snapshot store snapshot exceeds its request bound");
  let parsed: unknown;
  try { parsed = JSON.parse(value); } catch { throw new ArgumentError("HTTP snapshot store snapshot must be valid JSON"); }
  if (!isObject(parsed)) throw new ArgumentError("HTTP snapshot store snapshot must be a JSON object");
  if (canonicalJson(parsed) !== value) throw new ArgumentError("HTTP snapshot store snapshot must use canonical JSON");
  return value;
}

function responseError(operation: AutonomousHttpSnapshotStoreOperation, status: number): TransportError {
  const retryable = status === 408 || status === 425 || status === 429 || status >= 500;
  return new TransportError(`HTTP snapshot store ${operation} returned status ${status}${retryable ? " (retryable)" : ""}`);
}

/**
 * A portable remote text store for any strict metadata snapshot persistence adapter.
 *
 * The server contract is deliberately small: GET returns 200 with canonical JSON or 404 when
 * absent; PUT returns any 2xx on success; conditional PUT returns 409/412 for a CAS miss. The
 * expected digest is sent as an HTTP validator, never as a persisted payload field.
 */
export class AutonomousHttpSnapshotTextStore {
  readonly endpoint: URL;
  readonly resource: string;
  readonly requireHttps: boolean;
  readonly allowLoopback: boolean;
  readonly timeoutMs: number;
  readonly maxRequestBytes: number;
  readonly maxResponseBytes: number;
  readonly headerResolver?: AutonomousHttpSnapshotStoreHeaderResolver;
  readonly fetch: AutonomousHttpSnapshotStoreFetch;
  readonly signal?: AbortSignal;
  private readonly allowedHosts: readonly string[];

  constructor(options: AutonomousHttpSnapshotStoreOptions) {
    if (!options || typeof options !== "object") throw new ArgumentError("HTTP snapshot store options are malformed");
    this.requireHttps = options.requireHttps ?? true;
    this.allowLoopback = options.allowLoopback ?? false;
    if (typeof this.requireHttps !== "boolean" || typeof this.allowLoopback !== "boolean") throw new ArgumentError("HTTP snapshot store HTTPS policy must be boolean");
    this.allowedHosts = (options.allowedHosts ?? []).map(normalizeHost);
    if (this.allowedHosts.length > 128) throw new ArgumentError("HTTP snapshot store allowed hosts exceed their bound");
    this.endpoint = validateEndpoint(options.endpoint, this.allowedHosts, this.requireHttps, this.allowLoopback);
    this.resource = boundedText("HTTP snapshot store resource", options.resource, MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESOURCE_BYTES);
    if (!/^[A-Za-z0-9_.:/+-]+$/.test(this.resource)) throw new ArgumentError("HTTP snapshot store resource contains unsafe characters");
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    if (!Number.isSafeInteger(this.timeoutMs) || this.timeoutMs < 100 || this.timeoutMs > MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_TIMEOUT_MS) throw new ArgumentError("HTTP snapshot store timeoutMs is outside its bound");
    this.maxRequestBytes = options.maxRequestBytes ?? MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_REQUEST_BYTES;
    this.maxResponseBytes = options.maxResponseBytes ?? MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESPONSE_BYTES;
    if (!Number.isSafeInteger(this.maxRequestBytes) || this.maxRequestBytes < 1 || this.maxRequestBytes > MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_REQUEST_BYTES) throw new ArgumentError("HTTP snapshot store maxRequestBytes is outside its bound");
    if (!Number.isSafeInteger(this.maxResponseBytes) || this.maxResponseBytes < 1 || this.maxResponseBytes > MAX_AUTONOMOUS_HTTP_SNAPSHOT_STORE_RESPONSE_BYTES) throw new ArgumentError("HTTP snapshot store maxResponseBytes is outside its bound");
    if (options.fetch !== undefined && typeof options.fetch !== "function") throw new ArgumentError("HTTP snapshot store fetch must be callable");
    if (options.headerResolver !== undefined && typeof options.headerResolver !== "function") throw new ArgumentError("HTTP snapshot store headerResolver must be callable");
    const fetcher = options.fetch ?? globalThis.fetch;
    if (typeof fetcher !== "function") throw new ArgumentError("HTTP snapshot store requires a fetch implementation");
    this.fetch = options.fetch ?? fetcher.bind(globalThis);
    this.headerResolver = options.headerResolver;
    this.signal = options.signal;
  }

  describe(): AutonomousHttpSnapshotStoreDescription {
    return {
      schema: AUTONOMOUS_HTTP_SNAPSHOT_STORE_SCHEMA,
      resource: this.resource,
      host: this.endpoint.hostname,
      scheme: this.endpoint.protocol === "https:" ? "https" : "http",
      require_https: this.requireHttps,
      allow_loopback: this.allowLoopback,
      timeout_ms: this.timeoutMs,
      max_request_bytes: this.maxRequestBytes,
      max_response_bytes: this.maxResponseBytes,
      cas: "if_match_digest_or_if_none_match_star",
      credentials: "transient_header_resolver;never_returned",
      retention: "metadata_only;caller_schema_validation_required",
      secret_material: "never_returned",
    };
  }

  async read(): Promise<string | null> {
    const result = await this.request("read", "GET", null, null);
    if (result.status === 404) return null;
    if (result.status < 200 || result.status >= 300) throw responseError("read", result.status);
    return validateSnapshotText(result.body, this.maxResponseBytes);
  }

  async write(value: string): Promise<void> {
    validateSnapshotText(value, this.maxRequestBytes);
    const result = await this.request("write", "PUT", value, null);
    if (result.status < 200 || result.status >= 300) throw responseError("write", result.status);
  }

  async writeIfUnchanged(expectedSnapshotDigest: string | null, value: string): Promise<boolean> {
    assertDigest("HTTP snapshot store expected snapshot digest", expectedSnapshotDigest);
    validateSnapshotText(value, this.maxRequestBytes);
    const result = await this.request("write_if_unchanged", "PUT", value, expectedSnapshotDigest);
    if (result.status === 409 || result.status === 412) {
      return false;
    }
    if (result.status < 200 || result.status >= 300) throw responseError("write_if_unchanged", result.status);
    return true;
  }

  private async request(operation: AutonomousHttpSnapshotStoreOperation, method: "GET" | "PUT", body: string | null, expectedSnapshotDigest: string | null): Promise<{ status: number; body: string }> {
    if (!SAFE_METHODS.includes(method)) throw new ArgumentError("HTTP snapshot store method is unsupported");
    const context: AutonomousHttpSnapshotStoreHeaderContext = { operation, resource: this.resource, expected_snapshot_digest: expectedSnapshotDigest };
    let resolvedHeaders: Record<string, string> = {};
    if (this.headerResolver) {
      const value = await this.headerResolver(context);
      resolvedHeaders = normalizeHeaders(value);
    }
    const headers = new Headers({ accept: "application/json", ...resolvedHeaders });
    headers.set("x-aurora-snapshot-resource", this.resource);
    if (body !== null) {
      headers.set("content-type", "application/json");
      if (new TextEncoder().encode(body).byteLength > this.maxRequestBytes) throw new ArgumentError("HTTP snapshot store request exceeds its byte bound");
    }
    if (operation === "write_if_unchanged") {
      if (expectedSnapshotDigest === null) headers.set("if-none-match", "*");
      else headers.set("if-match", `\"${expectedSnapshotDigest}\"`);
    }
    const controller = new AbortController();
    let timedOut = false;
    const timer = setTimeout(() => {
      timedOut = true;
      controller.abort();
    }, this.timeoutMs);
    const abort = () => controller.abort();
    if (this.signal?.aborted) controller.abort();
    else this.signal?.addEventListener("abort", abort, { once: true });
    try {
      let response: Response;
      try {
        response = await this.fetch(this.endpoint, { method, headers, body, redirect: "error", signal: controller.signal });
      } catch (error) {
        if (timedOut) throw new TransportError(`HTTP snapshot store ${operation} timed out`);
        if (this.signal?.aborted) throw new TransportError(`HTTP snapshot store ${operation} was aborted`);
        throw new TransportError(`HTTP snapshot store ${operation} transport failed`, error);
      }
      if (!response || typeof response.status !== "number" || !response.headers) throw new ProtocolError("HTTP snapshot store returned a malformed response");
      let responseBody: string;
      try {
        responseBody = await boundedResponseText(response, this.maxResponseBytes);
      } catch (error) {
        if (error instanceof ResponseTooLargeError) throw error;
        throw new TransportError(`HTTP snapshot store ${operation} response could not be read`, error);
      }
      return { status: response.status, body: responseBody };
    } finally {
      clearTimeout(timer);
      this.signal?.removeEventListener("abort", abort);
    }
  }
}

export type AutonomousHttpSnapshotTextStoreDescription = AutonomousHttpSnapshotStoreDescription;
